src/Http/Request.php

```diff
@@ -0,0 +1,332 @@
+  const METHOD_POST = 'POST';
+  const METHOD_PUT = 'PUT';
+  const METHOD_PATCH = 'PATCH';
+  const METHOD_DELETE = 'DELETE';
+  const METHOD_PURGE = 'PURGE';
+  const METHOD_OPTIONS = 'OPTIONS';
+  const METHOD_TRACE = 'TRACE';
+  const METHOD_CONNECT = 'CONNECT';
+
+  /**
+   * HTTP Method.
+   *
+   * @var string
+   */
+  protected $method;
+
+  /**
+   * URI (path and query string).
+   *
+   * @var string
+   */
+  protected $uri;
+
+  /**
+   * Path
+   *
+   * @var string
+   */
+  protected $path;
+
+  /**
+   * Query parameters.
+   *
+   * @var array
+   */
+  protected $query;
+
+  /**
+   * The input HTTP headers.
+   *
+   * @var HttpHeaderBag
+   */
+  protected $headers;
+
+  /**
+   * The unprocessed body of the request.
+   *
+   * This should be a PHP stream, but let's keep it simple.
+   *
+   * @var string
+   */
+  protected $body;
+
+  /**
+   * Indicates if the request was routed by the menu system.
+   *
+   * @var bool
+   */
+  protected $viaRouter = FALSE;
+
+  /**
+   * The passed in CSRF token in the corresponding header.
+   *
+   * @var string
+   */
+  protected $csrfToken = NULL;
+
+  /**
+   * Cookies for the request.
+   *
+   * @var array
+   */
+  protected $cookies = array();
+
+  /**
+   * Files attached to the request.
+   *
+   * @var array
+   */
+  protected $files = array();
+
+  /**
+   * Server information.
+   *
+   * @var array
+   */
+  protected $server = array();
+
+  /**
+   * Holds the parsed body.
+   *
+   * @internal
+   * @var \ArrayObject
+   */
+  private $parsedBody;
+
+  /**
+   * Constructor.
+   *
+   * Parses the URL and the query params. It also uses input:// to get the body.
+   */
+  public function __construct($path, $query, $method = 'GET', HttpHeaderBag $headers, $via_router = FALSE, $csrf_token = NULL, $cookies = array(), $files = array(), $server = array()) {
+    $this->path = $path;
+    $this->query = $query;
+    $this->method = $method;
+    $this->headers = $headers;
+    $this->viaRouter = $via_router;
+    $this->csrfToken = $csrf_token;
+    $this->cookies = $cookies;
+    $this->files = $files;
+    $this->server = $server;
+
+    // Allow implementing modules to alter the request.
+    drupal_alter('restful_parse_request', $this);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create($path, $query, $method = 'GET', HttpHeaderBag $headers, $via_router = FALSE, $csrf_token = NULL, $cookies = array(), $files = array(), $server = array()) {
+    if ($method == static::METHOD_POST && $headers->get('x-http-method-override')) {
+      $method = $headers->get('x-http-method-override')->contents();
+    }
+    if (!static::isValidMethod($method, FALSE)) {
+      throw new \RestfulBadRequestException('Unrecognized HTTP method.');
+    }
+    return new static($path, $query, $method, $headers, $via_router, $csrf_token, $cookies, $files, $server);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createFromGlobals() {
```

Comment:
It's not called from anywhere yet. It will be called from the `RestfulManager` that will get instantiated on the menu callback.